### PR TITLE
Add all-in-one OTEL component

### LINKED ISSUE #2159: Exporter for In-Memory Storage

### LINKED ISSUE #2258: Create all-in-one OTEL distribution
Related to https://github.com/jaegertracing/jaeger/issues/2160 and https://github.com/jaegertracing/jaeger/issues/2159

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves #2258 
Resolves #2159 

Add all-in-one based on OpenTelemetry collector

Notable changes:
* OTEL in-memory exporter - `jaeger_memory`
* query service is configurable only via flags
* archive storage in query is configurable only via flags
* removed dependency of admin service from query. Now the query reports HC status changes via channel
* refactored default configuration 

Follow up issues/PR:
* expose OTEL logger https://github.com/open-telemetry/opentelemetry-collector/pull/1032
* subscribe to OTEL lifecycle events https://github.com/open-telemetry/opentelemetry-collector/issues/1033

### CODE REVIEW

File: cmd/opentelemetry-collector/app/defaults/default_config_test.go [hunk 1]
```diff
@@ -19,433 +19,154 @@ import (
...
-						},
+						Receivers: []string{kafkaRec.TypeStr},
+						Exporters: []string{elasticsearch.TypeStr},
 					},
 				},
 			},
 		},
 		{
-			args: []string{
-				"--processor.jaeger-binary.server-host-port=host:1",
```
Comment (objectiser) [inline cmd/opentelemetry-collector/app/defaults/default_config_test.go:337]: Looks like this test has been removed? I think this is still required.

File: cmd/opentelemetry-collector/app/defaults/default_config_test.go [hunk 2]
```diff
@@ -19,433 +19,154 @@ import (
...
-	"go.opentelemetry.io/collector/receiver"
-	"go.opentelemetry.io/collector/receiver/jaegerreceiver"
-	"go.opentelemetry.io/collector/receiver/zipkinreceiver"
 	"go.uber.org/zap"
 
 	"github.com/jaegertracing/jaeger/cmd/opentelemetry-collector/app"
 	"github.com/jaegertracing/jaeger/cmd/opentelemetry-collector/app/exporter/cassandra"
 	"github.com/jaegertracing/jaeger/cmd/opentelemetry-collector/app/exporter/elasticsearch"
 	"github.com/jaegertracing/jaeger/cmd/opentelemetry-collector/app/exporter/grpcplugin"
-	"github.com/jaegertracing/jaeger/cmd/opentelemetry-collector/app/exporter/kafka"
```
Comment (objectiser) [inline cmd/opentelemetry-collector/app/defaults/default_config_test.go:40]: Would be worth adding kafka exporter in one of the tests, just for coverage.

File: Makefile [hunk 3]
```diff
@@ -280,6 +280,14 @@ else
 	cd ${OTEL_COLLECTOR_DIR}/cmd/ingester && $(GOBUILD) -o ./opentelemetry-ingester-$(GOOS) $(BUILD_INFO) main.go
 endif
 
+.PHONY: build-otel-all-in-one
+build-otel-all-in-one:
+ifeq ($(GOARCH), s390x)
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply Makefile:285]: Yes there will be a conflict also in other parts of the makefile.

File: cmd/all-in-one/main.go [hunk 4]
```diff
@@ -224,7 +224,14 @@ func startQuery(
 ) *queryApp.Server {
 	spanReader = storageMetrics.NewReadMetricsDecorator(spanReader, baseFactory.Namespace(metrics.NSOptions{Name: "query"}))
 	qs := querysvc.NewQueryService(spanReader, depReader, *queryOpts)
-	server := queryApp.NewServer(svc, qs, qOpts, opentracing.GlobalTracer())
+	server := queryApp.NewServer(svc.Logger, qs, qOpts, opentracing.GlobalTracer())
+	go func() {
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/all-in-one/main.go:228]: my bad, copy-paste fatal error

File: cmd/opentelemetry-collector/app/defaults/default_config_test.go (hunk 2 above)
Comment (pavolloffay) [reply cmd/opentelemetry-collector/app/defaults/default_config_test.go:40]: It's already there but was using a string, I will change it to use the constant.

File: cmd/opentelemetry-collector/app/defaults/default_config.go [hunk 5]
```diff
@@ -74,11 +93,15 @@ func CollectorConfig(storageType string, zipkinHostPort string, factories config
 	}, nil
 }
 
-func createCollectorReceivers(zipkinHostPort string, factories config.Factories) configmodels.Receivers {
-	jaeger := factories.Receivers["jaeger"].CreateDefaultConfig().(*jaegerreceiver.Config)
-	if jaeger.Protocols == nil {
-		jaeger.Protocols = map[string]*receiver.SecureReceiverSettings{}
+func createReceivers(component ComponentType, zipkinHostPort string, factories config.Factories) configmodels.Receivers {
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/opentelemetry-collector/app/defaults/default_config.go:96]: I have done this on purpose - be explicit on the _required_ API contract.

[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/opentelemetry-collector/app/defaults/default_config.go:96]: The public API is easy to use. I like to keep the internal APIs clear in terms of the required arguments.

File: cmd/opentelemetry-collector/app/defaults/default_config_test.go (hunk 1 above)
Comment (pavolloffay) [reply cmd/opentelemetry-collector/app/defaults/default_config_test.go:337]: The code is fully covered in this class. The same test is in `jaeger_receiver_test.go`.

File: cmd/opentelemetry-collector/app/exporter/memory/factory_test.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	jConfig "github.com/jaegertracing/jaeger/pkg/config"
+)
+
+func TestCreateTraceExporter(t *testing.T) {
+	v, _ := jConfig.Viperize(AddFlags)
+	factory := &Factory{Viper: v}
+	exporter, err := factory.CreateTraceExporter(context.Background(), component.ExporterCreateParams{Logger: zap.NewNop()}, factory.CreateDefaultConfig())
+	require.NoError(t, err)
+	require.NotNil(t, exporter)
+}
```
Comment (objectiser) [inline cmd/opentelemetry-collector/app/exporter/memory/factory_test.go:39]: Should there be a test to ensure that the exporter is using a singleton?

File: cmd/opentelemetry-collector/cmd/all-in-one/main.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+	svc, err := service.New(service.Parameters{
+		ApplicationStartInfo: info,
+		Factories:            cmpts,
+		ConfigFactory:        cfgFactory,
+	})
+	handleErr(err)
+
+	// Add Jaeger specific flags to service command
+	// this passes flag values to viper.
+	storageFlags, err := app.AddStorageFlags(storageType, true)
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/opentelemetry-collector/cmd/all-in-one/main.go:114]: I will try to make the archive storage separate to the OTEL components.

File: cmd/opentelemetry-collector/app/exporter/memory/factory_test.go (hunk 6 above)
Comment (pavolloffay) [reply cmd/opentelemetry-collector/app/exporter/memory/factory_test.go:42]: +1 added

**Comment by pavolloffay** [comment]
@objectiser I have removed the archive storage config from exporters. It looks cleaner now.

File: cmd/opentelemetry-collector/cmd/all-in-one/main.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+		handleErr(fmt.Errorf("could not load Jaeger configuration file %w", err))
+	}
+
+	go func() {
+		err = svc.Start()
+		handleErr(err)
+	}()
+
+	configCreatedWait.Wait()
+	exp := getStorageExporter(storageType, exporters)
```
Comment (objectiser) [inline cmd/opentelemetry-collector/cmd/all-in-one/main.go:143]: What if a comma separated list of storage types are provided?

File: cmd/opentelemetry-collector/cmd/all-in-one/main.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+		archiveOpts.InitFromViper(v)
+		primaryConfig := exporter.(*cassandra.Config)
+		f := cassStorage.NewFactory()
+		f.InitFromOptions(cassStorage.NewOptionsFromConfig(primaryConfig.Primary.Configuration, archiveOpts.Primary.Configuration))
+		if err := f.Initialize(metrics.NullFactory, logger); err != nil {
+			return nil, err
+		}
+		return f, nil
+	case "memory":
+		return memory.GetFactory(), nil
```
Comment (objectiser) [inline cmd/opentelemetry-collector/cmd/all-in-one/main.go:210]: Why not badger, grpc_plugin?

Comment (pavolloffay) [reply cmd/opentelemetry-collector/cmd/all-in-one/main.go:222]: badger hasn't been implemented yet. I will add the `grpc-plugin`.

File: cmd/opentelemetry-collector/cmd/all-in-one/main.go (hunk 8 above)
Comment (pavolloffay) [reply cmd/opentelemetry-collector/cmd/all-in-one/main.go]: That does not seem like that somebody would use multiple storages with all-in-one. We can default to the first one in the list.

That said I would remove the support for the list that can be done via OTEL config.

**Comment by pavolloffay** [comment]
The last fixes building OTEL all-in-one image and runs integration test for it. 

I will merge on green.

**Comment by RehanSaeed** [comment]
Is there any documentation regarding this Docker image? In particular, which ports I need.

**Comment by objectiser** [comment]
@RehanSaeed The information is contained in this section: https://www.jaegertracing.io/docs/1.21/opentelemetry/

**Pull request merged**